### PR TITLE
USHIFT-936, USHIFT-942: reconcile the internal and external config data structures

### PR DESCRIPTION
**Pull request by dhellmann**
This PR eliminates some technical debt by merging the internal and
external representation of our configuration. The PR is structured with
several commits that should each work independently to make reviewing
the large number of changes simpler.

/assign @pmtk
/assign @pacevedom
/assign @copejon

### CODE REVIEW

**Comment by dhellmann** [comment]
Tests failing with 
```
Mar 04 00:23:38 release-ci-ci-op-swi8jhnb-5da1b microshift[212294]: ??? F0304 00:23:38.868198  212294 run.go:46] Error in reading or validating configuration: Failed to validate configuration file /etc/microshift/config.yaml: network.serviceNetwork not filled in
[96](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_microshift/1442/pull-ci-openshift-microshift-main-e2e-router-smoke-test/1631802154794618880#1:build-log.txt%3A96)
```

File: pkg/config/files.go
Comment (pliurh) [inline pkg/config/files.go:86]: This breaks the CI. This function is invoked in `fillDefaults`. And it will set `c.ApiServer.SkipInterface` to true when it's invoked again.

File: pkg/config/files.go
Comment (pliurh) [inline pkg/config/files.go:97]: Shall we invoke `incorporateUserSettings` after `fillDefaults`? So that user settings can always overwrite the defaults.

File: pkg/config/config.go
Comment (pliurh) [inline pkg/config/config.go:160]: If a user set `SkipInterface` to false? Shall we return an error here?

File: pkg/config/files.go
Comment (dhellmann) [reply pkg/config/files.go:97]: Yes, great point. The fillDefaults function tries to be smart, but it might be better to have that be simple and move the smarts to the other function. I will try to do that tomorrow.

File: pkg/config/config.go
Comment (dhellmann) [reply pkg/config/config.go:160]: The SkipInterface value is a computed value, so it is not currently being read from the file at all. c.userSettings.ApiServer.SkipInterface should always be false. 

@pacevedom is working on another PR that exposes the flag, although we haven't had time to talk about why that's necessary. https://github.com/openshift/microshift/pull/1478

File: pkg/config/files.go
Comment (dhellmann) [reply pkg/config/files.go]: Done

File: pkg/config/config.go
Comment (pacevedom) [inline pkg/config/config.go:58]: Should we use constants for the fixed values in this function?

**Review by pacevedom — APPROVED** [review]
Two nits, looking good.

File: pkg/config/etcd.go
Comment (dhellmann) [reply pkg/config/etcd.go:11]: I pushed an update for this.

File: pkg/config/config.go
Comment (dhellmann) [reply pkg/config/config.go:58]: We could. The main point of having constants is to reuse the same value in multiple places. In this case most of these values only appear the one time, except for tests. I could prepare another PR to see what the changes would look like.

Comment (pacevedom) [reply pkg/config/config.go:58]: That was my point, yes. In order to reduce cognitive load when changing tests (or even md docs in the repo), have them all in a single place easy to look for. I am ok with doing it in another PR.

**Comment by pacevedom** [comment]
/lgtm
/hold
I am ok with the change, holding just in case you want more reviews. Feel free to remove it.

File: Makefile
Comment (pmtk) [inline Makefile:247]: PTAL https://github.com/openshift/microshift/blob/main/scripts/auto-rebase/rebase.sh#L929
We can remove `(cd "${dirpath}" && make vendor)` and just run `make vendor` in `microshift/`

File: pkg/config/files.go
Comment (pmtk) [inline pkg/config/files.go:92]: 1st `if` looks unnecessary

**Comment by dhellmann** [comment]
@pacevedom @pmtk  I have rebased this and it is ready for another review.

File: Makefile
Comment (dhellmann) [reply Makefile:247]: I worry that just running `make vendor` will have other side effects for the top level vendor dir. I can change that line to run `make vendor-etcd` though.

Comment (dhellmann) [reply Makefile:247]: Actually, looking closer at that script, it updates each directory one at a time and commits each separately. We can't do that if we run `make vendor` and using `make vendor-etcd` assumes that there is only 1 submodule.

I think there's some work to optimize that script, but maybe we can leave it for another PR to give us time to consider how we want it to work?

File: Makefile
Comment (dhellmann) [reply Makefile:252]: I'll change it to use make recursively.

File: pkg/config/files.go
Comment (dhellmann) [reply pkg/config/files.go:92]: Good catch, thanks!

File: pkg/controllers/kube-apiserver.go
Comment (dhellmann) [reply pkg/controllers/kube-apiserver.go]: Good idea, done

File: Makefile
Comment (dhellmann) [reply Makefile]: I've removed this line for now. I'm trying to do is get to a point where we could run a job that tries to update all of the vendor directories and fails if any changes are found (because we want developers to update the vendor dir in their PR if they change something that needs to be revendored), but I can do that in another PR.

**Comment by pmtk** [comment]
/lgtm
/hold
@pacevedom please unhold when you review

**Comment by dhellmann** [comment]
I've marked this WIP because @dusk125 is going to be removing the etcd configuration options (more details to follow elsewhere). I will rebase this change on top of that work when it is done.

**Comment by dhellmann** [comment]
Rebased to include #1506

File: pkg/config/apiserver.go
Comment (eggfoobar) [inline pkg/config/apiserver.go:29]: If the `ApiServer.URL` and `Port` value is an internally created one and not changing at runtime, what do you think about changing the type to `URL` and populate a `Port` attribute on instantiation, that way we push out the error check so it's not on the caller to handle?

```go
type ApiServer struct {
        ...
	// The URL of the API server
	URL  *url.URL `json:"-"`
        // The Port of the API server
        Port int      `json:"-"`
}
```

File: pkg/config/debugging.go
Comment (eggfoobar) [inline pkg/config/debugging.go:12]: What do you think about doing a `strings.ToLower(c.Debugging.LogLevel)` and match on lowercase to avoid the odd chance a user adds incorrect capitalization?

Comment (dhellmann) [reply pkg/config/debugging.go:12]: That makes sense. This PR is already pretty big. Maybe we can do that in a follow-up?

File: pkg/config/apiserver.go
Comment (dhellmann) [reply pkg/config/apiserver.go:29]: I think I like that. I can experiment with it when I rebase this on #1506

File: pkg/config/debugging.go
Comment (eggfoobar) [reply pkg/config/debugging.go:12]: Yeah absolutely, I have no issue with it.

**Comment by dhellmann** [comment]
/hold cancel

I believe this is ready for review.

File: pkg/config/apiserver.go
Comment (dhellmann) [reply pkg/config/apiserver.go]: I decided to stick with a string for the URL, since that's how it's used, and added a separate field for the port.

File: pkg/config/debugging.go
Comment (dhellmann) [reply pkg/config/debugging.go:12]: I opened USHIFT-1043 to remember to do this later.

**Comment by eggfoobar** [comment]
LGTM, I was able to pull this and run it with out any issues.

File: pkg/config/files.go
Comment (pmtk) [inline pkg/config/files.go:44]: It just occurred to me that we wanted to remove user dir (microshift needs to be run as root anyway), but given the size of the PR and effort spent on reviewing it we might do it in another PR :)

**Comment by pmtk** [comment]
@eggfoobar gave +1. Also saw @copejon +1 on slack
/lgtm

**Pull request merged**
